### ISSUE TITLE
Fetch is not defined in firebase webpack

### ISSUE THREAD

**Issue by cwooi95** [firebase/firebase-js-sdk#4663]
### [REQUIRED] Describe your environment

  * Operating System version:  Mac OS Catalina
  * Browser version: N/A
  * Firebase SDK version:  "firebase": "^7.24.0",
  * Firebase Product:  (auth, storage, firestore, hosting)

### [REQUIRED] Describe the problem

When upgrading to latest firebase version and running angular universal build for ssr, it throws error of "ReferenceError: fetch is not defined" pointing to firebase webpack module. I have not used any fetch function in the server.ts, while it is throwing fetch error. (I am not using node server, plan to run angular project fully hosted on firebase)

Error:
<img width="952" alt="Fetch is not defined" src="https://user-images.githubusercontent.com/38025127/111933135-d16cf280-8af9-11eb-82a9-d94f8189230f.png">

Error log:
[2021-03-22T02_05_48_192Z-debug.log](https://github.com/firebase/firebase-js-sdk/files/6179087/2021-03-22T02_05_48_192Z-debug.log)

#### Steps to reproduce:

I am using these two lines in package.json after installing angular-universal.
    "serve:ssr": "node dist/server",
    "build:ssr": "npm run build:client-and-server-bundles && npm run compile:server",

#### Relevant Code:
server.ts
```javascript
import 'zone.js/dist/zone-node';
import 'reflect-metadata';
import {enableProdMode} from '@angular/core';
// Express Engine
import {ngExpressEngine} from '@nguniversal/express-engine';
// Import module map for lazy loading
import {provideModuleMap} from '@nguniversal/module-map-ngfactory-loader';

import express from 'express';
import {join} from 'path';

// To solve "ReferenceError: XMLHttpRequest is not defined"
// DOM libs required for Firebase
(global as any).WebSocket = require('ws');
(global as any).XMLHttpRequest = require('xhr2');

// Faster server renders w/ Prod mode (dev mode never needed)
enableProdMode();

// Express server
const app = express();

// Compress the file to gzip
const compression = require('compression');
app.use(compression());

const PORT = process.env.PORT || 4000;
const DIST_FOLDER = join(process.cwd(), 'dist/browser');

// * NOTE :: leave this as require() since this file is built Dynamically from webpack
const {AppServerModuleNgFactory, LAZY_MODULE_MAP} = require('./dist/server/main');

// Our Universal express-engine (found @ https://github.com/angular/universal/tree/master/modules/express-engine)
app.engine('html', ngExpressEngine({
  bootstrap: AppServerModuleNgFactory,
  providers: [
    provideModuleMap(LAZY_MODULE_MAP)
  ]
}));

app.set('view engine', 'html');
app.set('views', DIST_FOLDER);

// Example Express Rest API endpoints
// app.get('/api/**', (req, res) => { });
// Server static files from /browser
app.get('*.*', express.static(DIST_FOLDER, {
  maxAge: '1y'
}));

// All regular routes use the Universal engine
app.get('*', (req, res) => {
  res.render('index', { req });
});

// Start up the Node server
app.listen(PORT, () => {
  console.log(`Node Express server listening on http://localhost:${PORT}`);
});

```

webpack.server.config.js
```javascript
const path = require('path');
const webpack = require('webpack');

module.exports = {
  mode: 'none',
  entry: {
    // This is our Express server for Dynamic universal
    server: './server.ts'
  },
  target: 'node',
  resolve: {
    extensions: ['.ts', '.js']
   },
  optimization: {
    minimize: false
  },
  output: {
    // Put this to resolve "window is not defined"
    globalObject: "this",
    // Puts the output at the root of the dist folder
    path: path.join(__dirname, 'dist'),
    filename: '[name].js'
  },
  module: {
    rules: [
      { test: /\.ts$/, loader: 'ts-loader' },
      {
        // Mark files inside `@angular/core` as using SystemJS style dynamic imports.
        // Removing this will cause deprecation warnings to appear.
        test: /(\\|\/)@angular(\\|\/)core(\\|\/).+\.js$/,
        parser: { system: true },
      },
    ]
  },
  plugins: [
    new webpack.ContextReplacementPlugin(
      // fixes WARNING Critical dependency: the request of a dependency is an expression
      /(.+)?angular(\\|\/)core(.+)?/,
      path.join(__dirname, 'src'), // location of your src
      {} // a map of your routes
    ),
    new webpack.ContextReplacementPlugin(
      // fixes WARNING Critical dependency: the request of a dependency is an expression
      /(.+)?express(\\|\/)(.+)?/,
      path.join(__dirname, 'src'),
      {}
    )
  ]
};

```

**Comment by looptheloop88** [comment]
Hi @cwooi95, I'm having some difficulties in replicating this issue. Would you mind sharing a minimal, but with complete code implementation (e.g. including the package.json file) in a sample project that I can run locally to replicate this issue?

**Comment by cwooi95** [comment]
@looptheloop88 Apologise for the late reply, it took me some time to reproduce the error with minimal code, you can get the code from this public repo
[https://github.com/cwooi95/Minimum_reproducible_example](https://github.com/cwooi95/Minimum_reproducible_example)

Let me know if you need more info.

Thank you.

**Comment by looptheloop88** [comment]
Hi @cwooi95, thanks for sharing a sample project. This is probably a similar [issue](https://github.com/firebase/firebase-js-sdk/issues/1455#issuecomment-455712500) and here are the [options](https://github.com/firebase/firebase-js-sdk/issues/1797#issuecomment-498474850) that you can implement as a workaround. You'll need to check what's best suited for your case, because it will affect how dependencies are getting imported and referenced.

**Comment by cwooi95** [comment]
For option 1, when I apply alias to firebase and firebase/app, it throws error as below:
<img width="1249" alt="Screenshot 2021-03-25 at 11 11 32 PM" src="https://user-images.githubusercontent.com/38025127/112496145-779d4e80-8dbf-11eb-948c-cdb611f7d2d8.png">
After adding alias for auth, firestore, performance, I am still getting the same error.


I also tried with the option 2, and it successfully solves the fetch undefined problem, however it throws another error "XMLHttpRequest is not defined". So I have imported the xmlhttprequest and xhr in the server.ts file, but it still doesn't recognise XMLHttpRequest. Any workaround/suggestions for this error?

Below are the lines added in server.ts
```
(global as any).WebSocket = require('ws');
(global as any).XMLHttpRequest = require('xmlhttprequest').XMLHttpRequest;
```


Using option 3 still gives me the same error of "fetch is not defined".
